### DISCUSSION TITLE
Improve how WooCommerce displays images with a fixed aspect ratio

### DISCUSSION THREAD

**Discussion by Aljullu** [woocommerce/woocommerce#49343] [WooCommerce, Blocks, and Site Editing]
> [!NOTE]
> 
> * In some places you will see I mention images having a specific size but I only mention one number. That's the horizontal width of the image. WordPress uses the width of the image, independently of the height, to generate thumbnails.
> * In many places you will also see I speak about "cropped" images. With that I refer to images which are displayed in the frontend with a fixed aspect ratio, which might be different than the original aspect ratio of the image.

## The problem

You might be familiar with the [different sizes](https://developer.wordpress.org/themes/functionality/featured-images-post-thumbnails/#image-sizes) an image is resized to when it's uploaded to a WordPress site. Those are `thumbnail` (150px), `medium` (300px), `medium_large` (768px), `large` (1024px) and full size. They are useful so WordPress can use one thumbnail or the other depending on the context. In fact, thanks to HTML responsive images (the ones with `srcset` attribute), it might be the browser the one who decides which thumbnail to load.

That's beneficial to WooCommerce as well, any time we print an image to the screen, we can use one of those sizes or rely on `srcset` to handle it for us. There is only one caveat: those images are uncropped, and they keep the aspect ratio the original image has. This is an issue because most of the times WooCommerce needs to display cropped product images, which can be in a square aspect ratio or any other ratio specified in the _Customizer_ (`/wp-admin/customize.php`). When WooCommerce displays a cropped image, it can't use the thumbnails generated by WordPress, as they have a different aspect ratio, so it needs to use some other thumbnails generated by WooCommerce itself, those are `woocommerce_thumbnail` and `woocommerce_gallery_thumbnail` which are 100px and 300px pixels wide respectively.

This poses a big issue, if the biggest cropped image we have is 300px wide, that means that any time WooCommerce has to display a cropped image bigger than that, it will look blurry.

![image](https://github.com/woocommerce/woocommerce/assets/3616980/7ca46c1a-df19-4023-994b-9fa3f8182ddd)

_(If you open the image in a new tab or zoom into it, you will see _Blouse_ image is blurry, that's because its original aspect ratio was not a square, so WooCommerce can't use WordPress-generated thumbnails, instead it has to use `woocommerce_gallery_thumbnail`, which is 300px wide)_

## Suggested solutions

I could think of two solutions to this issue:

### Solution 1: Generate bigger cropped thumbnails

This is what I initially proposed in https://github.com/woocommerce/woocommerce/issues/28916#issuecomment-2162938417. With this approach, we would define some new thumbnail sizes for cropped images, so when rendering a cropped image, there are more thumbnails to choose from.

I took a quick look to see how different themes display product images: 

Theme | Single product gallery thumbnails | Shop page thumbnails
--- | --- | ---
TT4 | 128px | 412px
TT3 | 83px | 318px
TT2 | 83px | 318px
T17 | 80px | 205px
Storefront | 46px | 231px
Astra | 138px | 285px
Blocksy | 113px | 410px
GeneratePress | 88px | 227px
Kadence | 60px | 387px
OceanWP | 95px | 271px
Tsubaki | 83px | 318px

As you can see, 100px and 300px worked quite well in most themes when visited with a device with a low-resolution display. But with many devices nowadays having 2x density pixels and themes showing bigger images (TT4 displays _Shop_ page thumbnails at 412px by default), those sizes that worked well some years ago don't work that well nowadays. Another reason is that the _Product image_ block that can be found inside the _Product Collection_ or _Single Product_ blocks allows displaying product images even bigger than those sizes.

Based on all of that, it's clear that having the maximum thumbnail size of 300px is way to low. My suggestion would be to update and introduce new cropped thumbnail sizes. Those could be:

* 150px
* 300px
* 600px
* 1200px

Or, alternatively, we could use the same sizes as WordPress' uncropped thumbnails:

* 150px
* 300px
* 768px
* 1024px

**Pros:**

1. Having images cropped to the aspect ratio they are printed with in the screen is the best for performance. Only the pixels which will be visible in the screen will be downloaded to the shopper's device.
2. This improvement would benefit any existing template, including classic themes.

**Cons:**

1. It will require much more disk space for existing stores.
2. If we generate thumbnails for one and only one aspect ratio, what if we want to add more aspect ratio options to a block? The generated images won't benefit those blocks if a different aspect ratio is selected.

### Solution 2: Use uncropped images and crop them in the frontend with CSS (WordPress-way)

This idea came from looking at how WordPress handles selecting different aspect ratios in blocks like _Featured image_. The trick is that the image they use keeps the original aspect ratio, but via some CSS they make it look with the aspect ratio selected by the user.

![imatge](https://github.com/woocommerce/woocommerce/assets/3616980/d3dcb13e-eeb4-4aba-a821-93389d7ecfb0)

_(In the picture you can see the image is 800x800px, but it's displayed in a 620x348px form thanks to the usage of `object-fit: cover`)_

**Pros:**

1. It allows selecting different aspect ratios in any block, without having to generate thumbnails for each aspect ratio.
2. Doesn't use unnecessary disk space, as WooCommerce doesn't need to generate thumbnails for different aspect ratios.
3. Makes good use of the browser cache because no matter which aspect ratio is used to render the images, the same file will be used.

**Cons:**

1. Blocks that display images in a different aspect ratio than their original, end up loading a bigger image than what's rendered, as it will be cropped in one direction or the other. This might affect performance negatively, but stores can easily work it around uploading product images in the aspect ratio they want to display them in the frontend.
2. It doesn't solve the issue for classic themes or PHP-rendered templates.
3. Requires updating some existing blocks so they stop using the cropped images and instead use the uncropped ones with some custom CSS.

## Thoughts?

What are your thoughts on this? While I initially leaned towards solution 1, the more I think about it, the more I like solution 2.

The main reasons for me are:

1. Simplicity: we don't need to worry about generating WooCommerce-specific thumbnails anymore. We could even forget about the _Product images_ setting in the _Customizer_ and don't port it to block themes. Things will be simpler for themes as well, as they won't have to battle with WooCommerce thumbnail sizes and aspect ratios.
2. Flexibility: it allows us to add an _Aspect Ratio_ setting to any block without worrying about whether there will be a thumbnail that can be used.
3. Disk space: stores won't require additional disk space to store all new thumbnails.

It's true this solution doesn't address the issue for classic themes or PHP templates, but in most cases they work just fine. The instances we know where images are displayed with a lower resolution than expected are related to blocks.

**Comment by coreymckrill** [comment]
+1 for option 2. I like the idea of eliminating custom image sizes that take extra resources/time to generate every time you upload a new image. I think the cons can be worked around:

> Blocks that display images in a different aspect ratio than their original, end up loading a bigger image than what's rendered, as it will be cropped in one direction or the other. This might affect performance negatively, but stores can easily work it around uploading product images in the aspect ratio they want to display them in the frontend.

I think the impact would be relatively minor. Side question: do our blocks lazy load images the way that [WP Core does](https://make.wordpress.org/core/2020/07/14/lazy-loading-images-in-5-5/)?

> It doesn't solve the issue for classic themes or PHP-rendered templates.

Seems like the CSS for PHP-rendered templates could still be adjusted to accommodate this in most cases.

> Requires updating some existing blocks so they stop using the cropped images and instead use the uncropped ones with some custom CSS.

Would [block deprecation](https://developer.wordpress.org/block-editor/reference-guides/block-api/block-deprecation/) smooth out this process?

**Reply by Aljullu** [reply]
> I think the impact would be relatively minor. Side question: do our blocks lazy load images the way that [WP Core does](https://make.wordpress.org/core/2020/07/14/lazy-loading-images-in-5-5/)?

Good question! I just tested and images are lazy loaded in the Product Collection block, for example. If I'm understanding it right, WordPress is already taking care of it, so there is nothing special that blocks need to do. (I might be wrong, though :thinking: )

> Seems like the CSS for PHP-rendered templates could still be adjusted to accommodate this in most cases.

Yes, that's a good idea. I'm a bit wary when modifying existing PHP templates as existing stores and themes might be relying on the existing markup/CSS, but it's definitely worth exploring!

> Would [block deprecation](https://developer.wordpress.org/block-editor/reference-guides/block-api/block-deprecation/) smooth out this process?

Yeah, I didn't dive deeper into which blocks would need to be updated and how complex it would be. In some of them we would need to deprecate the block and use the block deprecation API, in others it might be even easier as they are printed dynamically with PHP code so it's just a matter of updating the render function.

**Comment by tjcafferkey** [comment]
I've spent the past day or so looking into and this is what I've found (apologies if some of this is repeated from the original post, its as much for my own understanding as it is for yours). For the context on the below I'll be using `woocommerce_thumbnail`.

### Removing custom WooCommerce image sizes.

The problem, as stated by @Aljullu is that fundamentally we need to render thumbnails in various places and at various sizes but we only have a 300x300 image as our largest option. This means that when rendered in an area bigger than that, the quality of the image degrades. This ultimately means we are going to need to load better, slightly larger images to improve the quality and experience resulting in:

* Larger payload to load more appropriate images.
* A variety of different sizes. Since we won't be using `woocommerce_thumbnail` the images won't always be guaranteed to be 300x300 any longer.

### Usages: The Product Collection block

The thumbnail is used in a variety of places, lets consider the Product Collection block for its flexibility

* Columns can vary from [2 columns](https://github.com/user-attachments/assets/a9b8df30-e7e8-4ae6-9418-65f9d5960934) to [6 columns](https://github.com/user-attachments/assets/f9856817-52aa-4c5a-9646-4a7e0c7e1a16) and even [a stack](https://github.com/user-attachments/assets/e4a7e2bf-418e-4745-9713-29feb1263b15), resulting in various different widths for the image to take up when being rendered.
* Can be rendered [full-width of the viewport](https://github.com/user-attachments/assets/4d8b4b3e-2286-472e-b6dc-2c1b6003d0ed). We shouldn't assume it will _always_ be used in a constrained Group block. Resulting in larger image widths.
* Currently renders square images by default regardless of original aspect ratio, so any changes to the markup and images will need to preserve this.

The block's flexibility creates numerous scenarios affecting image size and quality. While perfecting all cases is challenging, we should focus on optimizing defaults and improving edge cases. Currently, a 300x300 image is stretched across all uses, which we can significantly enhance.

### The Solution.

Native image `srcset` and `sizes` attributes can improve the experience. By leaning into WordPress' core functions [wp_calculate_image_srcset()](https://developer.wordpress.org/reference/functions/wp_calculate_image_srcset/) and [wp_calculate_image_sizes()](https://developer.wordpress.org/reference/functions/wp_calculate_image_sizes/) we can remove the maintenance burden and let core WP and browsers handle image optimisations.

#### Understanding srcset and sizes

srcset and sizes are HTML attributes that work together to serve responsive images:

* **srcset**: Lists image variations and their widths `<img srcset="beanie-2.jpg 801w, beanie-2-300x300.jpg 300w" ... />`
  * `801w` indicates the image is 801 pixels wide
  * `300w` indicates the image is 300 pixels wide
* **sizes**: Specifies image display size based on viewport `<img sizes="(max-width: 300px) 25vw, 300px" ... />`
  * For viewports up to 300px wide, image occupies 25% of viewport width
  * For larger viewports, image is 300px wide

The browser uses these attributes to select and load the most appropriate image for the current display conditions. These examples are from our _current_ HTML. Hopefully now you understand why we're getting 300x300 images.

#### Benefits

* Browsers can choose the most appropriate image size
* Improved image quality across various display scenarios
* Reduced maintenance burden by leveraging WordPress core & browser APIs
* A step forward to removing the need to crop images thus reducing disk space.

#### Backwards compatibility.

To ensure the images are still rendered in a square like they currently are we can look to use the CSS `aspect-ratio` property for this. @Aljullu has done a good job of explaining the how this works in the original post.

We may want to look at converting the [existing display options](https://github.com/user-attachments/assets/fc036b35-a8f3-4443-b1c8-cb415ca0b267) so they are more aligned with [Gutenbergs Feature Image options](https://github.com/user-attachments/assets/8819fc0c-7bf7-4f62-bd55-e87c58eb6acc) or at least replace "Cropped" with "Square - 1:1" to begin with and migrate those attributes using the [deprecation APIs migration method](https://developer.wordpress.org/block-editor/reference-guides/block-api/block-deprecation/)

**Comment by Aljullu** [comment]
> Solution 2: Use uncropped images and crop them in the frontend with CSS (WordPress-way)

I opened an issue for this here: https://github.com/woocommerce/woocommerce/issues/54241.

**Comment by Marc-pi** [comment]
@Aljullu is your solution the same as the one explained by @tjcafferkey ???
please take care of loading time on mobile, loading a big image is not appropriate
srcset / picture tag are the best options, and core does handle those ([wp_calculate_image_srcset()](https://developer.wordpress.org/reference/functions/wp_calculate_image_srcset/) and [wp_calculate_image_sizes()](https://developer.wordpress.org/reference/functions/wp_calculate_image_sizes/) 
>> see his PR https://github.com/woocommerce/woocommerce/pull/54192

**Reply by Aljullu** [reply]
My understanding is that they are two different discussions: https://github.com/woocommerce/woocommerce/pull/54192 is to use responsive images using a ResizeObserver while https://github.com/woocommerce/woocommerce/issues/54241 is to move away from WooCommerce-generated cropped thumbnails to the WordPres-generated uncropped thumbnails.

**Comment by yjailin** [comment]
Just ran into this on a vanilla install — product thumbnails are noticeably blurry on retina screens because the srcset only goes up to 300px. On WP.com this is handled automatically via Jetpack Photon, but self-hosted stores with no CDN or image optimization plugin get blurry thumbnails by default, with no way to fix it without a third-party plugin.

For merchants investing in good product photography, this is a real problem. It undermines the whole storefront experience before a customer even reads a product title.

<img width="1443" height="748" alt="Capture d’écran 2026-06-17 à 11 33 27" src="https://github.com/user-attachments/assets/58cbe2ee-604e-414d-8e02-83911c5cc7c0" />

Solution 2 makes sense to me. Happy to help test or contribute if this gets picked up.

**Comment by Aljullu** [comment]
Thanks for all the feedback, folks! I'm going to close this discussion as I opened an issue about implementing solution 2 here: https://github.com/woocommerce/woocommerce/issues/65963.
